Exclude invalid identifiers from __dir__. Include normal attrs.

danielballan opened the pull request:
This also adds a separate `__dir__` implementation for `RemoteCatalog`, as described in #235 and in tests and comments here.

Closes #235.

martindurant commented:
Any thoughts on defining `_ipython_key_completions_()` (see https://github.com/NCAR/intake-siphon/pull/2 ) ?

martindurant commented on intake/catalog/base.py:
```diff
@@ -364,6 +381,24 @@ def __init__(self, url, http_args={}, page_size=None, **kwargs):
 
+    def __dir__(self):
+        # Include (cached) tab-completable entries and normal attributes.
+        warnings.warn(
```
I think this should only appear in the case that we know there are more entries than we are going to show, otherwise could be quite annoying

danielballan replied:
Good point! Fixed and thoroughly tested.

danielballan commented:
Rebased on master, made the warnings appear only when applicable (needed one more piece of public state on `Entries` to make it possible cleanly), and implemented `_ipython_key_completions_`. Thanks for the suggestion!

martindurant commented:
Looks good, thank you!
